### PR TITLE
Add Annotations to GroupBy

### PR DESCRIPTION
**Pull request by WillAyd**
Not fully complete but got started on these on a flight today. Should be able to review failures in a few more details.

@gwrome @vaibhavhrt would love any feedback from you

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1980,8 +2017,12 @@ def cumcount(self, ascending=True):
...
     @Substitution(name='groupby')
     @Appender(_common_see_also)
-    def rank(self, method='average', ascending=True, na_option='keep',
-             pct=False, axis=0):
+    def rank(self,
+             method: str = 'average',
+             ascending: bool = True,
+             na_option: str = 'keep',
+             pct: bool = False,
+             axis=Axis) -> Union[Series, DataFrame]:
```
Comment (vaibhavhrt) [inline pandas/core/groupby/groupby.py]: I think it should be `axis: Axis = 0`

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -1916,11 +1952,12 @@ def ngroup(self, ascending=True):
             index = self._selected_obj.index
             result = Series(self.grouper.group_info[0], index)
             if not ascending:
-                result = self.ngroups - 1 - result
+                result = self.ngroups - result
```
Comment (gwrome) [inline pandas/core/groupby/groupby.py]: This change is breaking `pandas/tests/groupby/test_counting.py::TestCounting`

**Review by gwrome — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: Yea thanks that was a mistake changing in the first place; this generates a mypy I was trying to debug but haven't seen all of the way through yet

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/groupby/groupby.py]: Nice catch!

**Comment by WillAyd** [comment]
Will come back to this at a later date

**Pull request closed without merging**